Copter: (WIP) Calculate time and MAH required for RTL

Pedals2Paddles opened the pull request:
WIP - Soliciting thoughts and advice...

- [X] Calculate time and MAH required for each phase of flight during RTL.  Utilizes various copter speed parameters, 4 new parameters that define the MAH rate in each phase of flight, the copter position, and home position.  This provides a calculated _Total MAH to RTL_.
- [x] Add a parameter which allows the operator to specify the desired battery percentage to be at upon landing.  For example, I want to always be on the ground with 5% remaining. The battery monitor can then take 5% of the total capacity, and add that to the calculated _Total MAH to RTL_.
- [ ] Use the _percent remaining at landing_ plus the _Total MAH to RTL_ values as a live dynamic battery MAH remaining alarm/failsafe trigger.  This would be used instead of the fixed battery MAH remaining parameter used today.
- [ ] Integrate with Smart RTL, calculating based on that cruising distance instead of the straight line RTL distance
- [ ] Something to enable/disable this. Trying to avoid a million parameters, but will need something.
- [ ] Automatically learn the MAH parameters similar to how we automatically detect the hover throttle.

New parameters as of 8/5/2018:
- `rtl_mah_climb`
- `rtl_mah_cruise`
- `rtl_mah_hover`
- `rtl_mah_descent`

Pedals2Paddles commented:
As of opening this WIP PR, I have the mathematics actually working nicely.  I've performed multiple RTls from various distances and altitudes.  In each case, the predicted time and MAH verified with the actual time and energy.  I filled in the MAH parameters for phase of flight manually based on real world observation and log analysis of what my copter utilizes.  My Solo happens to consume battery as follows:
- `rtl_mah_climb` = 25
- `rtl_mah_cruise` = 19
- `rtl_mah_hover` = 21
- `rtl_mah_descent` = 15

Since the reworking of the battery monitor, I don't see any existing way to send the total MAH required to the battery monitor. @peterbarker?

SergeyBokhantsev commented:
@Pedals2Paddles great feature, some ideas here from my side
- energyToRTL() function need to detect a case when some mandatory setting(s) not set (zeroes). So to not to provide wrong calculations.
- rtl_mah_hover could be a single parameter to enable/disable the feature. If other parameters leaved zeros (by default) then they are could be roughly calculated from rtl_mah_hover as: 
rtl_mah_climb = rtl_mah_hover + 25%
rtl_mah_cruise = rtl_mah_hover
rtl_mah_descent = rtl_mah_hover - 20%
- so the learning feature could only learn the rtl_mah_hover value
- (not sure) maybe rtl_mah_climb, cruise and descend parameters units should be the percents from rtl_mah_hover?
- user could set rtl_mah_climb, cruise and descend parameters explicitly to achieve best accuracy.

BaptisteEspivent commented:
Hi Matt, I have been following your previous PR and was glad to see that you were working on an update! I have used your code as base and added mah estimation as you were suggesting in your last bullet point. I tested it and it looks to be working quite fine and after a few RTLs, it quickly converged towards a steady mah estimation for each phase.
Problem I had was to estimate actual capacity of the battery. Have you been working on this? I implemented a function that sets remaining capacity at startup base on correspondance table between pack voltage and theorical percentage left. This works quite fine but depending on how old the battery is, it doesn't really deliver its full theorical capacity which is problematic to estimate capacity left...
Some other features i thought about 

- take into account wind direction and wind speed if known (from external source or from estimation)

-  Extend consumption estimation to auto mode and be able to predict if vehicle will be able to complete the mission with the remaining battery

Pedals2Paddles commented:
@SergeyBokhantsev I'm trying to minimize new parameters, so I would indeed like to use one of these as the enable/disable parameter too.  As an alternative, I could have it remain disabled unless all are filled out.  I thought about the +/- percentage, but I think it would be too generic and only accurate for a few copters of similar design.  I would prefer not to use fuzzy math on a failsafe.

@BaptisteEspivent What method did you use to learn the MAH?  Regarding battery capacity... I can see no reliable and repeatable way to determine the battery's real capacity.  We just can't programatically know what random battery someone has plugged in.  This only works for smart batteries (like the Solo), and the code already automatically reads and updates the battery capacity.  This is again something that I can't use fuzzy math or assumptions on since it is a failsafe. I have to be able to say "it will do exactly this when you do that."  That leads to operator entered parameters, and it will do exactly what the operator told it do.

BaptisteEspivent commented:
I used the same logic as for throttle_hover estimation : I added a parameter to enable/disable energy estimation learning. I also added a function that will update the different power estimations. If estimation learning is set to LEARN or LEARN_AND_SAVE, my update function will be called in each state of the mode_rtl.cpp and update corresponding parameter (climb, cruise, hover, land). The update function is coded as for throttle_hover estimation : 
hover_power = hover_power + (dt/(dt+AP_ENERGY_ESTIMATOR_POWER_TC))*(watts() - hover_power);

liang-tang commented:
nice work, really useful feature.

Pedals2Paddles commented:
So this is now calculating the total required mAh to perform an RTL based on the user-entered mAh rates for each phase of flight.  _To do: write an automated learn and save function_.   There is a parameter for the user to specify they want to be on the ground at X% remaining.  So it adds that % of the total battery capacity mAh + the mAh required to perform an RTL to calculate a new live failsafe MAH value.

Added a function in the battery monitor for that value to be ingested.  And that is now checked in the battery monitor's failsafe check in the same manner the fixed mAh failsafe is checked.

If it isn't raining, I'll try to test fly this evening to verify it acts as expected.

Personally, I would like to see this take precedence over the existing fixed mAh failsafe using a MIN() function.  This would allow it to dynamically maximize flying time by minimizing failsafe mAh.

Pedals2Paddles commented:
- Added learn and save of hover current in the same manner that hover throttle is learned and saved.
- Modified new parameters for climb, cruise, and descent power to be a % +/- hover current.
- Added checking of `BATT_FS_LOW_ACT` to see what the operator has selected as a failsafe.
  - If failsafe action is RTL, the full mah required to RTL is sent to the battery monitor.
  - If failsafe action is LAND, only the landing phase mah required is sent to the battery monitor.
  - If failsafe action is disabled, smartRTL, or terminate, calculation is not used and 0 is sent to the battery monitor.
- MAH required to RTL is now added to the existing `FS_BATT_LOW_MAH` parameter. Much simpler to implement and much simpler for the user.   When using the RTL energy calculation, the failsafe will trigger at `FS_BATT_LOW_MAH` + RTL MAH, making the copter touch down right at the `FS_BATT_LOW_MAH` point.  If you are not using the RTL energy calculation, the failsafe will trigger right at `FS_BATT_LOW_MAH`, just like it always has.
- Major cleanup of variables

I see a few errors and clean up still needed now that I'm awake.  More testing and tweaking this evening.

Pedals2Paddles commented:
Updated the original post to reflect current tasks and methods.
